R/sum-duplicate-genes.R

```diff
@@ -0,0 +1,102 @@
+#' Sum counts for genes with duplicate names in a SingleCellExperiment object.
+#'
+#' Genes with the same name are merged by summing their raw expression counts.
+#' If requested, the lognormalized expression values are recalculated, otherwise
+#' this is left blank.
+#'
+#'
+#' @param sce a SingleCellExperiment object with duplicated row names
+#' @param normalize a logical indicating whether to normalize the expression
+#'   values. Default is TRUE
+#' @param recalculate_reduced_dims a logical indicating whether to recalculate
+#'   PCA and UMAP. If FALSE, the input reduced dimensions are copied over. If
+#'   TRUE, the highly variable genes are also recalculated with the new values
+#'   stored in metadata. Default is FALSE
+#'
+#' @return a SingleCellExperiment object
+#' @export
+#'
+#' @import SingleCellExperiment
+#' @import SummarizedExperiment
+#'
+sum_duplicate_genes <- function(sce, normalize = TRUE, recalculate_reduced_dims = FALSE) {
+  stopifnot(
+    "sce must be a SingleCellExperiment object" = is(sce, "SingleCellExperiment"),
+    "normalize must be a logical" = is.logical(normalize),
+    "recalculate_reduced_dims must be a logical" = is.logical(recalculate_reduced_dims),
+    "normalize can not be FALSE if recalculate_reduced_dims is TRUE" = normalize || !recalculate_reduced_dims
+  )
+  if (normalize) {
+    stopifnot(
+      "Package `scran` must be installed if `normalize = TRUE` is set." =
+        requireNamespace("scran", quietly = TRUE),
+      "Package `scuttle` must be installed if `normalize = TRUE` is set." =
+        requireNamespace("scuttle", quietly = TRUE)
```

Comment:
Same comment as I left for `Seurat` and `SeuratObject`, but the more I look the more I prefer the clarity of 2 separate statements.

Reply:
same response: `scran` depends on `scuttle` so the check is mostly redundant.